ProcessLib/TwoPhaseFlowWithPrho/TwoPhaseFlowWithPrhoMaterialProperties.h

```diff
@@ -0,0 +1,250 @@
+        double const beta =
+            (drhoh2nonwet_dpg - drhoh2wet_dpg) *
+            PG;  // NOTE here should be PG^h, but we ignore vapor
+        double const dPC_dSw =
+            _capillary_pressure_models[_current_material_id]->getdPcdS(S);
+        return alpha / (beta - alpha * dPC_dSw);
+    }
+    /// Calculate the derivative using the analytical way
+    double calculatedSwdX(double const PL, const double /*X*/, const double S,
+                          const double rho_wet_h2,
+                          double const T = 303.15) const
+    {
+        const double PG = PL +
+                          _capillary_pressure_models[_current_material_id]
+                              ->getCapillaryPressure(S);
+        double const rho_equilibrium_wet_h2 = PG * Henry_const * molar_mass_h2;
+        if ((1 - S) < (rho_equilibrium_wet_h2 - rho_wet_h2))
+        {
+            return 0.0;
+        }
+        double const drhoh2wet_dpg = Henry_const * molar_mass_h2;
+        double const drhoh2nonwet_dpg = molar_mass_h2 / R / T;
+        double const alpha =
+            ((drhoh2nonwet_dpg - drhoh2wet_dpg) * (1 - S) + drhoh2wet_dpg);
+        double const beta =
+            (drhoh2nonwet_dpg - drhoh2wet_dpg) *
+            PG;  // NOTE here should be PG^h, but we ignore vapor
+        double const dPC_dSw =
+            _capillary_pressure_models[_current_material_id]->getdPcdS(S);
+        return -1 / (beta - alpha * dPC_dSw);
+    }
+    /// Calculate the derivative using the analytical way
+    double calculatedXmdX(double PL, double Sw, double rho_wet_h2,
+                          double dSwdX) const
+    {
+        const double PG = PL +
+                          _capillary_pressure_models[_current_material_id]
+                              ->getCapillaryPressure(Sw);
+        double const rho_equilibrium_wet_h2 = PG * Henry_const * molar_mass_h2;
+        double const dPC_dSw =
+            _capillary_pressure_models[_current_material_id]->getdPcdS(Sw);
+        if ((1 - Sw) < (rho_equilibrium_wet_h2 - rho_wet_h2))
+            return 1.0;
+        return Henry_const * molar_mass_h2 * dPC_dSw * dSwdX;
+    }
+    /// Calculate the derivative using the analytical way
+    double calculatedXmdP(double PL, double Sw, double rho_wet_h2,
+                          double dSwdP) const
+    {
+        const double PG = PL +
+                          _capillary_pressure_models[_current_material_id]
+                              ->getCapillaryPressure(Sw);
+        double const rho_equilibrium_wet_h2 = PG * Henry_const * molar_mass_h2;
+        double const dPC_dSw =
+            _capillary_pressure_models[_current_material_id]->getdPcdS(Sw);
+        if ((1 - Sw) < (rho_equilibrium_wet_h2 - rho_wet_h2))
+            return 0.0;
+        return Henry_const * molar_mass_h2 * (1 + dPC_dSw * dSwdP);
+    }
+
+private:
+    double const Henry_const = 7.65e-6;  // mol/Pa./m3
```

Comment:
This Henry constant value comes from Henry law, which should belong to fluid property. Since currently this has not been implemented in fluid property yet, I will keep this here at this point. I will add a reference.